### PR TITLE
Exclude java.util.function from classes allowed in public API

### PR DESCRIPTION
**Pull request by ov7a**
### Context
Classes like Spec and Transformer from Gradle API should be used instead.

See also:
* #26770 
Test failure is expected until it's merged

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiCorrectnessTest.java [hunk 1]
```diff
@@ -70,7 +71,11 @@ public class PublicApiCorrectnessTest {
                 .or(type(BigDecimal.class))
                 .or(type(Element.class))
                 .or(type(QName.class))
-                .as("built-in JDK classes"))
+                .and(
+                    DescribedPredicate.not(resideInAnyPackage("java.util.function"))
+                        .or(type(BiFunction.class))
+                )
+                .as("built-in JDK classes (excluding java.util.function)"))
```
Comment (wolfs) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiCorrectnessTest.java:78]:
```suggestion
                .as("built-in JDK classes"))
```

💅 I'd keep the name here.

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiCorrectnessTest.java [hunk 2]
```diff
@@ -70,7 +71,11 @@ public class PublicApiCorrectnessTest {
                 .or(type(BigDecimal.class))
                 .or(type(Element.class))
                 .or(type(QName.class))
-                .as("built-in JDK classes"))
+                .and(
+                    DescribedPredicate.not(resideInAnyPackage("java.util.function"))
+                        .or(type(BiFunction.class))
+                )
```
Comment (wolfs) [inline subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiCorrectnessTest.java:77]: ❌ Let's not do this, but remove `java.util.function` from line 66. And add some comment that we don't want to have `java.util.function` including the reasons why.

**Review by wolfs — CHANGES_REQUESTED** [review]

File: subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiCorrectnessTest.java (hunk 2 above)
Comment (ov7a) [reply subprojects/architecture-test/src/test/java/org/gradle/architecture/test/PublicApiCorrectnessTest.java:77]: Thanks, missed that it's already ther

**Review by wolfs — APPROVED** [review]

**Comment by ov7a** [comment]
@bot-gradle test and merge

**Pull request merged**